examples: star height

davisjam opened the pull request:
Add an implementation of the "star height" heuristic
to detect regexes vulnerable to REDOS.

See also [https://github.com/substack/safe-regex](safe-regex).

Towards #154.

davisjam commented:
@DmitrySoshnikov Not quite in the style of the rest of the module. Let me know your thoughts.

DmitrySoshnikov commented on examples/count-star-height.js:
```diff
@@ -0,0 +1,31 @@
+	const starHeight = countStarHeight(obj.pattern, obj.options);
+	obj.starHeight = starHeight;
+}
+catch (e) {
```
You can install Prettier for auto-formatting (I'll open an issue with the default config file for it). Overall, JS style is:

```js
try {
  ...
} catch (e) {
  ...
}
```

davisjam replied:
I'll try out that tool.

DmitrySoshnikov commented on examples/count-star-height.js:
```diff
@@ -0,0 +1,31 @@
+	obj.starHeight = 'UNKNOWN';
+}
+
+console.log(JSON.stringify(obj));
```
```js
JSON.stringify(obj, null, 2);
```

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+ */
+
+const createRegExp = require('./create-regexp'),
+      regexpTree = require('../../');
```
ditto, please use standard style in the repo

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+/* Author: Jamie Davis <davisjam@vt.edu>
```
ditto, please use standard doc header with license, and copyright (and you can keep your name if you'd like)

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+function countStarHeight(regex, options) {
+	/* Get an ast. */
+	const re = createRegExp(regex);
+	console.error(`re: /${re.source}/`);
```
why is it `console.error`?

can use `console.info(...)` for normal messages

davisjam replied:
For my use of this example I wanted stdout to only contain "the answer". Any progress updates were to stderr to make it easier for me to script around the program. I can switch to `console.info` though.

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+	const ast = regexpTree.parse(re);
+
+	/* Options. */
+	const default_countQuestionMarks = false;
```
JS style, camelCase: `defaultCountQuestionMarks`

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+
+	/* Options. */
+	const default_countQuestionMarks = false;
+	const default_minimumRepetitionUpperLimit = 0;
```
ditto

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+	if (!options.minimumRepetitionUpperLimit) {
+		options.minimumRepetitionUpperLimit = default_minimumRepetitionUpperLimit;
+	}
+	console.error(`countQuestionMarks ${options.countQuestionMarks} minimumRepetitionUpperLimit ${options.minimumRepetitionUpperLimit}`);
```
`console.info`

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+	let maxObservedStarHeight = 0;
+
+	regexpTree.traverse(ast, {
+		'Repetition': {
```
no quotes just `Repetition:`

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+				if (node.quantifier && node.quantifier.kind === '?' && !options.countQuestionMarks) {
+					return;
+				}
+				else if (node.quantifier.kind === 'Range' &&
```
on previous line

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,83 @@
+				if (node.quantifier && node.quantifier.kind === '?' && !options.countQuestionMarks) {
+					return;
+				}
+				else if (node.quantifier.kind === 'Range' &&
```
```js
if (...) {
  ...
} else if (...) {
  ...
}
```

DmitrySoshnikov commented on examples/lib/create-regexp.js:
```diff
@@ -0,0 +1,75 @@
+const regexpTree = require('../../');
```
Use standard doc block comment with license and copyright.

DmitrySoshnikov commented on examples/lib/create-regexp.js:
```diff
@@ -0,0 +1,75 @@
+	});
+
+	// Remove comments.
+	pattern = pattern.replace(/\(\?#.*?\)/g, function(str, group1) {
```
use arrow function 

```js
..., (str, group1) => {
```

DmitrySoshnikov commented on examples/lib/create-regexp.js:
```diff
@@ -0,0 +1,75 @@
+	try {
+		reOrPattern = new RegExp(pattern);
+	}
+	catch (e) {
```
style

davisjam commented on examples/count-star-height.js:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env node
```
If you want to run it as `./count-star-height.js` without `node ./count-star-height.js` they are. For my purposes this was preferable, but if you would prefer the example need be invoked with the Node interpreter then I can switch it.

davisjam commented on examples/count-star-height.js:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env node
+/* Author: Jamie Davis <davisjam@vt.edu>
```
OK

davisjam commented:
Thoughts on including the linter as part of your git commit and push hooks? I assumed that was already in place.

DmitrySoshnikov commented:
Yeah, just an FYI, in https://github.com/DmitrySoshnikov/regexp-tree/commit/611adc23103665bdbbfc292982368bb4268afcca and https://github.com/DmitrySoshnikov/regexp-tree/commit/140938592b3f28ab1f05c1c8879764963fa83d88 I have setup Prettier. So if you rebase, do `npm install`, and `npm run build`, the linter, and Prettier formatter should run as a pre-commit hook. You can also setup the [pre-commit](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/scripts/git-pre-commit) hook manually.

davisjam commented:
@DmitrySoshnikov I believe I've addressed your review.


As I noted above, I'm leery about giving you copyright since this example is excerpted from my current research project. Unless I misunderstand what "copyright" means in this context?

DmitrySoshnikov commented on examples/lib/count-star-height.js:
```diff
@@ -0,0 +1,85 @@
+        if (node.quantifier && node.quantifier.kind === '?' && !options.countQuestionMarks) {
+          return;
+        } else if (node.quantifier.kind === 'Range' &&
+                   (!('to' in node.quantifier) || node.quantifier.to < options.minimumRepetitionUpperLimit))
```
Let's run Prettier nevertheless, it'll reformat it in the single style.

DmitrySoshnikov commented on examples/README.md:
```diff
@@ -0,0 +1,13 @@
+
+| File                   | Author                                               | Description       |
+|------------------------|------------------------------------------------------|-------------------|
+| `count-star-height.js` | Jamie Davis [@davisjam](https://github.com/davisjam) | Correct implementation of the well-known and buggy [safe-regex](https://github.com/substack/safe-regex) project. |
```
OK, here what I though.

This actually might be a good separate project by itself. Maybe you should create a separate repo for this at all instead of having it in the main regexp-tree repo? With this you fully preserve your own copyright.

Since the example is specific, it might be less practical as specifically an example for a quick start to use the tool. However, I can see it as a good example of already using the tool in other good projects like your research project.

davisjam replied:
OK, so there are really two discussions here:

1. Whether you're interested in `examples/` (and if so, what kind?)
2. Copyright.

I'm certainly not trying to start a fight about this, I just want to make sure we're on the same page.

# To `examples/` or not to `examples/`

> This actually might be a good separate project by itself.

Once the paper is accepted, I will be publishing my research repo to GitHub. The repo includes  this use of `regexp-tree` and a few others, plus a bunch of other unrelated code.

However, I'm loath to just dump my code on the internet and forget about it. I thought this would be a good place to put the `regexp-tree`-specific scripts as a way of giving back to the project.

We could link to a separate project with examples in the README, but then you end up with a link for every person who wants to share examples. Centralization seems better to me in this case.

> Since the example is specific

Of course it's specific :-). The entire `regexp-tree` project is pretty niche. Awesome, but niche.

> it might be less practical as specifically an example for a quick start to use the tool.

There are already many "quick-start" examples in the main README that demonstrate how to use individual features. I view the purpose of `examples/` as to show more complex uses of `regexp-tree` so people can get a sense of how they might want to use it themselves.

Did you have a different kind of code in mind for `examples/`?

# Copyright

> With this you fully preserve your own copyright.

If you want the examples it seems like we can address this with appropriate amendments to the copyright line in a few files.

DmitrySoshnikov replied:
@davisjam, yeah, that's correct, I'd like regexp-tree to maintain base core functionality, don't overload it with "concrete sub-projects" (like your research project), and avoid explicitly promoting/depending on other projects with direct links in README, etc.

All the contributors to this specific repo are explicitly mentioned in the https://github.com/DmitrySoshnikov/regexp-tree/graphs/contributors, with their explicit commits, etc, so their "copyright" is automatically "preserved" (assuming the MIT License of this project).

With this, the `examples\` directory I think should be completely independent from other projects (even as an "example dependency"), less focus on individual names, pages, etc, and more focus on actual examples which can help start using regexp-tree for some needs. These examples, as it seems to me, can be in general still very simple, but taking some smaller "real-world" issues.

For the specific research project, I really think it'll be better to open a separate repo. This will be much better for it, instead of putting it into some generic repo (which regexp-tree is).

davisjam commented:
Acknowledged. I've petitioned npm for publish rights to [safe-regex](https://github.com/substack/safe-regex) and will be putting this code there.

DmitrySoshnikov commented:
Sounds good, thanks!
